### ISSUE TITLE
Can the library api be improved to fit better with Elm?

### ISSUE THREAD

**Issue by jarlg** [bmatcuk/elm-webaudio#5]
Hey @bmatcuk, 

Eariler this summer I tried implementing [this gist](https://gist.github.com/giraj/250decbbc50ce091f79e) using elm-usermedia and elm-webaudio, but I was stumped by runtime errors I couldn't figure out. It's just a loop that analyses microphone input, picks out the loudest frequency at a given moment, and plays back an oscillator node of that frequency e.g you can control the oscillator (or potentially any synthetic sound) with your voice or whistle.

The errors seemed to relate quite deeply to Elm itself, and none of the libraries used. (Though I'll have to come back to this.) Nevertheless, using elm-webaudio felt a bit alien in Elm, i.e. not very functional nor reactive. It's true that the WebAudio API is neither of these things, and is indeed inherently mutable and everything not-Elm, but I'm hoping we could improve our interface with it. For my next attempt, I think I would access the WebAudio API through interop with javascript, to see how that feels, which is a nice way of saying I didn't particularly enjoy working with this library in Elm. (Please excuse my frankness -- it's a great effort! And I hope to build on it, but I also believe we could improve.)

I came across [this gist](https://gist.github.com/TheSeamau5/78adb0d3482445e22270) on the mailing lists, and it struck me as a potentially better way to model the WebAudio API in Elm. I'm not sure if you've already considered or seen this approach, I'd be interested to hear any thoughts and if you're open to a discussion on how the design of elm-webaudio should be?

Recently, tasks have been replaced by effects and I haven't been able to stay up-to-date on everything, but I'm guessing they could be used to improve on the gist above.

**Comment by bmatcuk** [comment]
Here's the thing: when I authored this library, Tasks didn't exist, nor did these "effects" things you mentioned. There was no nice way to implement a functional wrapper around a very not-functional API such as this. I did my best, and it worked for the time.

Since then, quite honestly, I've grown to dislike elm; I don't have the desire to keep up with the language changes. I think we should consider this project abandoned in favor of an implementation that utilizes newer language constructs. I'll note the abandoned status of this project in the README.

**Comment by jarlg** [comment]
@bmatcuk: Okay, totally understandable! Hope my critique wasn't too harsh -- I really didn't mean for it to be. The language has changed a lot indeed, and I think you've done good and important work opening up the possibility of interfacing with WebAudio, which it seems many haven't thought about (which is beyond me, since we have graphics and animation in Elm).

**Comment by bmatcuk** [comment]
No, not at all! You're right: it's pretty terrible now. But, back when I wrote it, the language had functions, lists, records, and unions and that was about it, so I had to make do =) I agree that a solid WebAudio library would compliment the graphics and animation stuff elm has, but it should be written by someone who has the time and desire to keep up with the language changes. I'm sure someone will take up the mantle... maybe you? =)
